NMS-14620: Quick Start - Configure Notifications (#5342)
OpenNMS/opennms: docs/modules/operation/pages/quick-start/notification-config.adoc

= Step 8: Configure Notifications

This section describes how to configure and customize your {page-component-title} notifications.

NOTE: This step is separate from xref:operation:quick-start/notifications.adoc[setting up default notifications].
We assume that you have followed the previous steps and are ready to proceed to more in-depth work with notifications.

== Configuration and customization

> Now that you've xref:operation:quick-start/baseline.adoc[established your environment's baseline], it's time to take your notes and configure your custom notifications.

Reviewer: ```suggestion
Now that you have xref:operation:quick-start/baseline.adoc[established your environment's baseline], it's time to take your notes and configure your custom notifications.
```

> Establishing notifications that are tailored for your {page-component-title} environment will help maintain it.

Reviewer: I think it's more that the notifications will help them maintain their network, rather than Horizon.
```suggestion
Establishing notifications that are tailored for your environment and monitoring requirements lets the right people see and respond to the right issues in a timely manner.
```

> Done correctly, customizing notifications will also mean that more relevant information is being brought to your attention.

Reviewer: ```suggestion

```

> Here, we will discuss how to configure event notifications and path outages.

Reviewer: ```suggestion
Here, we will describe how to configure event notifications and path outages.
```

Event notifications can be configured to provide UEI-specific coverage, allowing for granular control of incoming information.
Path outages prevent the generation of excess notifications by identifying network paths where failures may cause multiple nodes to appear offline.

== Configure event notifications

An event notification defines a destination path and message template for a specific event UEI.

> It is possible to define multiple notifications for the same UEI to utilize different filters, destination paths, and message templates.

Reviewer: ```suggestion
It is possible to define multiple notifications for the same UEI to use different filters, destination paths, and message templates.
```

> Configure event notifications by following these steps:

Reviewer: ```suggestion
Follow these steps to configure event notifications:
```

. Click the *Gear* symbol in the top-right of the screen.
. Under Event Management, click menu:Configure Notifications[Configure Event Notifications].
A list of events that are configured for notifications is displayed.
. Click *Add New Event Notification*, choose the UEI that will trigger a notification, and click *Next*.
** (Optional) Build a rule to match a subset of IP addresses or services, and click menu:Validate Rule Results[Next] or *Skip Results Validation*.

> . In the *Choose Path* screen, specify information to be included in the notification, including the destination path, subject (for email messages), body text, and so on.

Reviewer: ```suggestion
. In the *Choose Path* screen, specify information to include in the notification, including the destination path, subject (for email messages), body text, and so on.
```

Reviewer: Let's suggest they choose the destination path they defined earlier in the QS.

. Click *Finish*.

NOTE: Event notification configuration is stored in `$\{OPENNMS_HOME}/etc/notifications.xml`.
Changes to this file happen immediately, with no need to restart {page-component-title}.

=== Filter rules

> The filter rule configured in `$\{OPENNMS_HOME}/etc/notifications.xml` (for example, `<rule>IPADDR != '0.0.0.0'</rule>`) is not strict by default.

Reviewer: I'm not clear on what the filter rule is.

Author: From what I understand, it's a filter on notifications determining what to send and what to save. I'd like to have someone else verify that, though.

Reviewer 2: Reading further on into the doc, I agree that this rule, and the other version of the rule with the strict setting, determine what notifications to save and not to save. The wording is a little obtuse and makes some assumptions about the user fully understanding the context and could maybe use some examples as to what would trigger these different types of conditions for clarification? Maybe @mmahacek could provide some examples if he agrees?

Reviewer: Reading this again, I think if we add more of a definition, that should help. For example:

```suggestion
A filter rule lets you define which notifications to send or save. Strict rules are always evaluated, and if there is no node or interface associated with the event, the notification will not be saved. 

By default, the filter rule configured in `$\{OPENNMS_HOME}/etc/notifications.xml` (for example, `<rule>IPADDR != '0.0.0.0'</rule>`) is not strict.
That means that events with no associated node or interface will not validate the rule, and the notification would be saved.
```

Establishing notifications that are tailored for your environment and monitoring requirements lets the right people see and respond to the right issues in a timely manner.

Here, we will describe how to configure event notifications and path outages.
Event notifications can be configured to provide UEI-specific coverage, allowing for granular control of incoming information.
Path outages prevent the generation of excess notifications by identifying network paths where failures may cause multiple nodes to appear offline.

== Configure event notifications

An event notification defines a destination path and message template for a specific event UEI.
It is possible to define multiple notifications for the same UEI to use different filters, destination paths, and message templates.
Follow these steps to configure event notifications:

. Click the *Gear* symbol in the top-right of the screen.
. Under Event Management, click menu:Configure Notifications[Configure Event Notifications].
A list of events that are configured for notifications is displayed.
. Click *Add New Event Notification*, choose the UEI that will trigger a notification, and click *Next*.

> ** (Optional) Build a rule to match a subset of IP addresses or services, and click menu:Validate Rule Results[Next] or *Skip Results Validation*.

Reviewer 3: I think this needs to be called out as a step... while it is optional, the flow has the user going to the "Build Rule" page and selecting those IPs and Services.  I feel like this should be a step like the others but noted as being optional and how the rule will apply if they opt not to change anything.

Author: I personally tend to like moving optional steps into a sub-step rather than a numbered one, but it's definitely not a standard thing; I'll discuss it with Bonnie when she gets back, and change this step to a proper numbered one for now.

As for how the rule will apply should the user not make those optional changes, I'll need to do a quick bit of research and then update. :)

. In the *Choose Path* screen, specify information to include in the notification, including the destination path, subject (for email messages), body text, and so on.
. Click *Finish*.

NOTE: Event notification configuration is stored in `$\{OPENNMS_HOME}/etc/notifications.xml`.
Changes to this file happen immediately, with no need to restart {page-component-title}.

> === Filter rules

Reviewer 3: It's not entirely clear to me how this relates back to the rest of this section.  I think it would be good to include something about how this is related to the steps just taken in the web UI and how the two relate.

> A filter rule lets you define which notifications to send or save. Strict rules are always evaluated, and if there is no node or interface associated with the event, the notification will not be saved. 

Author: ```suggestion
A filter rule lets you define which notifications to send or save.
Strict rules are always evaluated, and if there is no node or interface associated with the event, the notification will not be saved.
Filter rules work alongside your event notification configuration to enable better, more fine-tuned control over the notifications that you receive.
```

By default, the filter rule configured in `$\{OPENNMS_HOME}/etc/notifications.xml` (for example, `<rule>IPADDR != '0.0.0.0'</rule>`) does not have the `strict` flag enabled.
That means if there is an event that is not associated with a node or interface (and therefore no `IPADDR` value), it will bypass validation and the notification would be processed.

A strict rule (for example, `<rule strict="true">IPADDR != '0.0.0.0'</rule>`) will always be evaluated, and if there is no node or interface associated with the event, the notification will not be saved.

== Configure path outages

Configuring a path outage lets you suppress notifications for nodes that appear to be down due to a failure in the network path between the nodes and {page-component-title}.
For example, if a WAN link fails, all nodes at the remote site served by the WAN link will appear to be down.
Because you get a notification that the router on the far end of the WAN link is not responding, you don't need notifications for all devices that sit behind that router.

> You can configure a path outage for an individual node by setting a parent on the node in a requisition, or create a rule-based configuration for a group of nodes:

Reviewer 3: This sentence sounds like we are explaining two options available to the user but then we go on to only list out the steps for one.  I feel like this should be listed as another thing a user can do or it could be explained as beyond the scope of this guide and focus the end user on one path.  It's also possible I am misreading this sentence.

Reviewer 4: Great place to link to a deep dive on how to configure it either way.

Author: It looks like we don't have a deep dive on this subject that goes into more detail; I actually lifted these instructions directly from the "Configure path outages" section of the Getting Started page. I think expanding on this in the Deep Dive will need to be a separate ticket; for now, I can remove the second part of the sentence.

```suggestion
You can configure a path outage for an individual node by setting a parent on the node in a requisition:
```

Follow these steps to configure event notifications:

. Click the *Gear* symbol in the top-right of the screen.
. Under Event Management, click menu:Configure Notifications[Configure Event Notifications].
A list of events that are configured for notifications is displayed.
. Click *Add New Event Notification*, choose the UEI that will trigger a notification, and click *Next*.
** (Optional) Build a rule to match a subset of IP addresses or services, and click menu:Validate Rule Results[Next] or *Skip Results Validation*.
. In the *Choose Path* screen, specify information to include in the notification, including the destination path, subject (for email messages), body text, and so on.
. Click *Finish*.

NOTE: Event notification configuration is stored in `$\{OPENNMS_HOME}/etc/notifications.xml`.
Changes to this file happen immediately, with no need to restart {page-component-title}.

=== Filter rules

A filter rule lets you define which notifications to send or save. Strict rules are always evaluated, and if there is no node or interface associated with the event, the notification will not be saved. 

> By default, the filter rule configured in `$\{OPENNMS_HOME}/etc/notifications.xml` (for example, `<rule>IPADDR != '0.0.0.0'</rule>`) is not strict.
> That means that events with no associated node or interface will not validate the rule, and the notification would be saved.
> That means if there is an event that is not associated with a node or interface, it will not validate the rule and, by default, the notification would be saved.

Reviewer 4: ```suggestion
By default, the filter rule configured in `$\{OPENNMS_HOME}/etc/notifications.xml` (for example, `<rule>IPADDR != '0.0.0.0'</rule>`) does not have the `strict` flag enabled.
That means if there is an event that is not associated with a node or interface (and therefore no `IPADDR` value), it will bypass validation and the notification would be processed.
```
